pkg/build/generator/generator.go

```diff
@@ -409,70 +405,70 @@ func (g *BuildGenerator) generateBuildFromConfig(ctx kapi.Context, bc *buildapi.
 			},
 		},
 		ObjectMeta: kapi.ObjectMeta{
+			Name:   buildName,
 			Labels: bcCopy.Labels,
 		},
 		Status: buildapi.BuildStatus{
 			Phase: buildapi.BuildPhaseNew,
 			Config: &kapi.ObjectReference{
 				Kind:      "BuildConfig",
-				Name:      bc.Name,
-				Namespace: bc.Namespace,
+				Name:      bcCopy.Name,
+				Namespace: bcCopy.Namespace,
 			},
 		},
 	}
-	if binary != nil {
-		build.Spec.Source.Git = nil
-		build.Spec.Source.Binary = binary
-		if build.Spec.Source.Dockerfile != nil && binary.AsFile == "Dockerfile" {
-			build.Spec.Source.Dockerfile = nil
-		}
-	} else {
-		// must explicitly set this because we copied the source values from the buildconfig.
-		build.Spec.Source.Binary = nil
-	}
 
-	build.Name = getNextBuildName(bc)
-	if build.Annotations == nil {
-		build.Annotations = make(map[string]string)
+	setBuildSource(binary, build)
+	setBuildAnnotationAndLabel(bcCopy, build)
+
+	var builderSecrets []kapi.Secret
+	var err error
+	if builderSecrets, err = g.FetchServiceAccountSecrets(bcCopy.Namespace, serviceAccount); err != nil {
+		return nil, err
 	}
-	build.Annotations[buildapi.BuildNumberAnnotation] = strconv.FormatInt(bc.Status.LastVersion, 10)
-	build.Annotations[buildapi.BuildConfigAnnotation] = bcCopy.Name
-	if build.Labels == nil {
-		build.Labels = make(map[string]string)
+	g.setBuildPushSecret(ctx, builderSecrets, bcCopy, &build.Spec.Output)
+	// Resolve image source if present
+	if err = g.setBuildSourceImage(ctx, builderSecrets, bcCopy, &build.Spec.Source); err != nil {
+		return nil, err
 	}
-	build.Labels[buildapi.BuildConfigLabelDeprecated] = buildapi.LabelValue(bcCopy.Name)
-	build.Labels[buildapi.BuildConfigLabel] = buildapi.LabelValue(bcCopy.Name)
-	build.Labels[buildapi.BuildRunPolicyLabel] = string(bc.Spec.RunPolicy)
-
-	builderSecrets, err := g.FetchServiceAccountSecrets(bc.Namespace, serviceAccount)
-	if err != nil {
+	if err = g.setBuildStrategy(ctx, builderSecrets, bcCopy, &build.Spec.Strategy); err != nil {
 		return nil, err
 	}
-	if build.Spec.Output.PushSecret == nil {
-		build.Spec.Output.PushSecret = g.resolveImageSecret(ctx, builderSecrets, build.Spec.Output.To, bc.Namespace)
+
+	return build, nil
+}
+
+// setBuildPushSecret set push secret for new build
+func (g *BuildGenerator) setBuildPushSecret(ctx kapi.Context, builderSecrets []kapi.Secret, bcCopy *buildapi.BuildConfig, outPut *buildapi.BuildOutput) {
+	if outPut.PushSecret == nil {
+		outPut.PushSecret = g.resolveImageSecret(ctx, builderSecrets, outPut.To, bcCopy.Namespace)
```

Comment:
this is another case where i think it would make more sense to make this a first class function that takes the secret as an argument, to make it a generally useful helper function.

also "outPut" should be "output"